dispersion_protocol/microservice.md

```diff
@@ -0,0 +1,171 @@
+| :-------------------------------------------------------------------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [MAV_CMD_REQUEST_MESSAGE](https://mavlink.io/en/messages/common.html#MAV_CMD_REQUEST_MESSAGE) | Request the target system(s) emit a single instance of a specified message. This is used to request [DISPERSION_DEVICE_INFORMATION](#dispersion_device_information). |
+
+| Enum                                                              | Description                                                                                                                          |
+| :---------------------------------------------------------------- | :----------------------------------------------------------------------------------------------------------------------------------- |
+| [DISPERSION_DEVICE_CAP_FLAGS](#dispersion_device_cap_flags)       | Dispersion device low level capability flags (bitmap). Used in [DISPERSION_DEVICE_INFORMATION](#dispersion_device_information).      |
+| [DISPERSION_DEVICE_STATUS_FLAGS](#dispersion_device_status_flags) | Flags for dispersion device (low level) operation (bitmap). Used in [DISPERSION_DEVICE_STATUS](#dispersion_device_status).           |
+| [DISPERSION_DEVICE_REQUESTS](#dispersion_device_requests)         | Dispersion device (low level) requests. Used in [DISPERSION_DEVICE_REQUEST](#dispersion_device_request).                             |
+| [DISPERSION_DEVICE_ERRORS](#dispersion_device_errors)             | Dispersion device (low level) error flags (bitmap, 0 means no error). Used in [DISPERSION_DEVICE_STATUS](#dispersion_device_status). |
+
+## Sequences
+
+Depicted below are message sequences for common scenarios.
+
+### Discovery
+
+The following diagram shows a possible sequence on startup that enables a ground station to discover a dispersion device and get its associated information.
+
+![discovery sequence diagram](discovery_sequence.png)
+
+### Control / Configuration
+
+Control of the dispersion system should be event driven, though the dispersion device status should be both at a regular cadence as well as driven by key events such as dispersion on and dispersion off. This sequence applies to requesting dispersion on and off as well as manually configuring the dispersion device before starting an automated mission. The user has the ability to swap out nozzles fairly regularly so the configuration step is needed to handle nozzle specific configuration data. Each request message has a request type and optional parameters to support the different control and configuration needs.
+
+![manual control sequence diagram](control_sequence.png)
+
+### Autopilot Control
```

Comment:
I would call this Mission control or similar.

A few considerations:

- A mav_cmd sent in a commands get a target system/component id, but these are not present in mission items. So if you think you'll ever have multiple dispersion devices you want to control then you will need to add a field dispersion_id to the command - see gimbal id in the gimbal messages. This allows you to specify the component id that you want the mission to resend the mission item "as a command" to. We also specify 1-6 to mean "send to an autopilot connected device".
- You will need to make sure the mission processing can track when the spray initiated by the mission command is complete - i.e. so that the system can proceed to the next step. This is likely to be a messy bit of integration.

Reply:
If there can really only be one dispersion device ever then this should be stated in the doc. The implication is that the autopilot will need to be configured to address commands to the locally connected sprayer OR to discover the connected MAVLink sprayer based on its type.

If you need to be able to sent the information in a mission it MUST be a command. It is OK to have multiple configuration commands, as long as we can group them sensibly.